index.bs

```diff
@@ -3205,7 +3442,7 @@ The following abstract operations support the implementation of the
  1. Let |firstDescriptor| be |controller|.[=ReadableByteStreamController/[[pendingPullIntos]]=][0].
  1. If |firstDescriptor|'s [=pull-into descriptor/byte offset=] + |firstDescriptor|' [=pull-into
     descriptor/bytes filled=] is not |view|.\[[ByteOffset]], throw a {{RangeError}} exception.
- 1. If |firstDescriptor|'s [=pull-into descriptor/byte length=] is not |view|.\[[ByteLength]], throw
+ 1. If |firstDescriptor|'s [=pull-into descriptor/byte length=] < |view|.\[[ByteLength]], throw
```

Comment:
I believe this was a missing feature of `byobRequest.respondWithNewView(view)`. You could already call `byobRequest.respondWith(bytesWritten)` with `0 <= bytesWritten <= byobRequest.view.byteLength`, but for some reason `.respondWithNewView(view)` **insisted** that you fill the entire view. Now, you can pass any view whose size is less than or equal to the original view's size. 🙂

Or would you prefer fixing this in a separate PR?

Reply:
Sorry for missing this. Doing this in the same PR makes sense, but I'm wondering if maybe we had a reason for this in the first place...

In particular, consider a consumer which wants to allocate and use, say, 1 MiB of memory. It allocates a 1 MiB buffer, and passes it through the stream, and gets it back, round and round.

My understanding is that in the current spec, no matter what, they will always get back a 1 MiB buffer. Even if the underlying byte source calls `byobRequest.respond(5)`, the returned buffer will still be 1 MiB (even if the _view_ is only 5 bytes). So when they cycle that buffer into another `reader.read(newViewOntoSameBuffer)` call, they can get up to 1 MiB that next time.

Whereas, if we allow smaller results to be sent back from `respondWithNewView()`, then the consumer would lose their 1 MiB buffer forever, and only get back a 5 byte buffer in return.

Now, that said, having written that up, it seems like we could still allow smaller views onto larger buffers. I.e. we could change the check from checking the view size to checking the buffer size. But I think we probably don't want to allow respondWithNewView() to take arbitrarily-small buffers?

Reply:
I agree. `byobRequest.respondWithNewView(newView)` should not allow passing a view whose buffer is *smaller* than the original buffer of the original BYOB request.

I would go even further: I'd say `byobRequest.respondWithNewView(newView)` must only be called with a view on the **same buffer** that was originally passed as `byobRequest.view.buffer`, after it has been **transferred** zero or more times. All of the data that was originally inside `byobRequest.view.buffer` should also be in `newView.buffer` (except for the part spanned by `newView`'s start and end, since that will contain the read data).

Ideally, I would make `.respondWithNewView()` enforce this: if `newView.buffer` is not the "transferred version" of `byobRequest.view.buffer`, then it should throw an error. Unfortunately, I don't think the ECMAScript specification provides any way to check this... 😞

We could approximate it by storing the original `byobRequest.view.buffer.byteLength` as part of the pull-into descriptor, and then checking whether `newView.buffer.byteLength` equals that original length. However, that doesn't prevent someone from creating a new `ArrayBuffer` with the same size and using that for `newView`... 😕

---

Either way, you **must** be able to pass a smaller view onto that buffer. Otherwise, it's impossible to satisfy the assert in step 3.1 of [ReadableByteStreamControllerRespondInternal](https://streams.spec.whatwg.org/commit-snapshots/dd76a17a3738d78708a8dfd8f0508e717d6a1988/#readable-byte-stream-controller-respond-internal) when called from `.respondWithNewView()`. And [we definitely need this](https://github.com/whatwg/streams/pull/1114#discussion_r601890120)! 😁

Reply:
Note that [this example in the specification](https://streams.spec.whatwg.org/#example-manual-read-bytes) *relies* on `read()` returning a view on the same (transferred) buffer. Otherwise the `offset < buffer.byteLength` check wouldn't work, and you might read more or fewer bytes than initially requested. It also assumes that data from previous reads is still available in `buffer`, so the underlying byte source cannot change any of the bytes outside of `byobRequest.view` either.

To summarize, the caller of `byobRequest.respondWithNewView(newView)` (i.e. the underlying byte source) must guarantee the following:
* `newView.byteOffset` must equal `byobRequest.view.byteOffset`, and `newView.byteLength` must be `<= byobRequest.view.byteLength`.
* `newView.buffer` has the exact same `byteLength` as `byobRequest.view.buffer`. This can be done by either transferring the original buffer, or by copying it (although that would defeat the purpose of using a BYOB request).
* Within `newView.buffer`, the byte ranges from `0` to `byobRequest.view.byteOffset` and from `byobRequest.view.byteOffset + byobRequest.view.byteLength` to `byobRequest.view.buffer.byteLength` must be unmodified. In other words, only bytes within the range of `byobRequest.view` can be changed. (This also applies to `byobRequest.respond(bytesWritten)`.)

Currently, we check the first condition. We could *feasibly* check the second one, by remembering the original `byteLength` as part of the pull-into descriptor. However, it's not possible to check the third condition without copying the original buffer upfront, which would defeat the purpose of BYOB. So we have to *assume* that the developer doesn't do anything weird in their underlying byte source... 😅

Reply:
> Currently, we check the first condition.

Well, _currently_ we kind of check the second one (by checking the view length, which must be >= the buffer length).

I'm thinking now we should pull out this change (to check both 1 + 2 cleanly), including the domintro change you suggest (which is a great idea), into a separate PR. Ideally we'd also add some usage examples for `respondWithNewView()`; I was surprised to not be able to find any in the spec.

Reply:
Solved in #1123.